chore: create NoopSpan instead reusing NOOP_SPAN

Flarna opened the pull request:
Always create a fresh `NoopSpan` instead of returning the same `NOOP_SPAN` several times.

Users may attach some data to a span or use it as a key in a map.
Avoid issues/leaks in such cases by always returning a new `NoopSpan`.

Flarna commented:
Added the `breaking` label because `NOOP_SPAN` was exported on API level.

johnbley commented:
> Added the `breaking` label because `NOOP_SPAN` was exported on API level.

Suppose we had already shipped 1.0 when you thought of this.  I assume that we would
- Document a warning to the effect mentioned above (e.g., "be aware that this is a singleton and may be a bad choice for a key")
- Change all our usage sites to `new NoopSpan()` as here
- Mark `NOOP_SPAN` as `@deprecated` and leave it in
I'm just treating this as a thought experiment for how we will handle this in the future; I think it's fine to break this tiny corner of the public api now.

dyladan commented:
> > Added the `breaking` label because `NOOP_SPAN` was exported on API level.
> 
> Suppose we had already shipped 1.0 when you thought of this. I assume that we would
> 
> * Document a warning to the effect mentioned above (e.g., "be aware that this is a singleton and may be a bad choice for a key")
> * Change all our usage sites to `new NoopSpan()` as here
> * Mark `NOOP_SPAN` as `@deprecated` and leave it in
>   I'm just treating this as a thought experiment for how we will handle this in the future; I think it's fine to break this tiny corner of the public api now.

This seems like a reasonable course of action

Flarna commented:
The missing part is to remove it on next major. Which in turn leads to questions like how often a new major will be made and if there will be some LTS support for older majors like in e.g. node.js. But that's a bigger discussion to be done outside of this PR.

But it shows once more that issues like #1765 and #1750 are needed to be handled before GA.

dyladan commented:
>how often a new major will be made

Current thinking of the spec sig is "very rarely." last I heard was "years"

Flarna commented:
Besides a major change in spec there could be JS local details causing majors. e.g. move to new typescript versions resulting in incompatibility with older ones,... de-support NodeJs 8,... But sure, I don't expect we reach v16 soon :o)

MSNev commented:
Suggestion: 
Rather than force the usage of ```new NoopSpan()``` everywhere (which can't be minified), change to same factory pattern like 

```
let noopSpan = createNoopSpan();

/**
 * The NoopSpan is the default {@link Span} that is used when no Span
 * implementation is available. All operations are no-op including context
 * propagation.
 */
export function createNoopSpan(): Span {
  let context: SpanContext = INVALID_SPAN_CONTEXT;
  let noopSpan:Span = {
    // Returns a SpanContext.
    context: () => context,

    // By default does nothing
    setAttribute: (_key: string, _value: unknown) => noopSpan,

    // By default does nothing
    setAttributes: (_attributes: SpanAttributes) => noopSpan,

    // By default does nothing
    addEvent: (_name: string, _attributes?: SpanAttributes) => noopSpan,

    // By default does nothing
    setStatus: (_status: SpanStatus) => noopSpan,

    // By default does nothing
    updateName: (_name: string) => noopSpan,

    // By default does nothing
    end: (_endTime?: TimeInput) => {},

    // isRecording always returns false for noopSpan.
    isRecording: () => false,

    // By default does nothing
    recordException: (_exception: Exception, _time?: TimeInput) => {}
  };

  return noopSpan;
}
```

This to both explicit as it's creating a new span, but we can then hide the details within the helper and the NoopSpan class code just disappears -- further assisting minification.

And size wise this becomes (could be optimized a little more)
```
function createNoopSpan() {
    var context = spancontext_utils_1.INVALID_SPAN_CONTEXT;
    var noopSpan = {
        // Returns a SpanContext.
        context: function () { return context; },
        // By default does nothing
        setAttribute: function (_key, _value) { return noopSpan; },
        // By default does nothing
        setAttributes: function (_attributes) { return noopSpan; },
        // By default does nothing
        addEvent: function (_name, _attributes) { return noopSpan; },
        // By default does nothing
        setStatus: function (_status) { return noopSpan; },
        // By default does nothing
        updateName: function (_name) { return noopSpan; },
        // By default does nothing
        end: function (_endTime) { },
        // isRecording always returns false for noopSpan.
        isRecording: function () { return false; },
        // By default does nothing
        recordException: function (_exception, _time) { }
    };
    return noopSpan;
}
exports.createNoopSpan = createNoopSpan;
```

vs the following where all of the ```NoopSpan.prototype.funcname``` are not minified.

```
var NoopSpan = /** @class */ (function () {
    function NoopSpan(_spanContext) {
        if (_spanContext === void 0) { _spanContext = spancontext_utils_1.INVALID_SPAN_CONTEXT; }
        this._spanContext = _spanContext;
    }
    // Returns a SpanContext.
    NoopSpan.prototype.context = function () {
        return this._spanContext;
    };
    // By default does nothing
    NoopSpan.prototype.setAttribute = function (_key, _value) {
        return this;
    };
    // By default does nothing
    NoopSpan.prototype.setAttributes = function (_attributes) {
        return this;
    };
    // By default does nothing
    NoopSpan.prototype.addEvent = function (_name, _attributes) {
        return this;
    };
    // By default does nothing
    NoopSpan.prototype.setStatus = function (_status) {
        return this;
    };
    // By default does nothing
    NoopSpan.prototype.updateName = function (_name) {
        return this;
    };
    // By default does nothing
    NoopSpan.prototype.end = function (_endTime) { };
    // isRecording always returns false for noopSpan.
    NoopSpan.prototype.isRecording = function () {
        return false;
    };
    // By default does nothing
    NoopSpan.prototype.recordException = function (_exception, _time) { };
    return NoopSpan;
}());
exports.NoopSpan = NoopSpan;
```

dyladan commented:
> Suggestion:
> Rather than force the usage of `new NoopSpan()` everywhere (which can't be minified), change to same factory pattern like
> 
> ```
> let noopSpan = createNoopSpan();
> 
> /**
>  * The NoopSpan is the default {@link Span} that is used when no Span
>  * implementation is available. All operations are no-op including context
>  * propagation.
>  */
> export function createNoopSpan(): Span {
>   let context: SpanContext = INVALID_SPAN_CONTEXT;
>   let noopSpan:Span = {
>     // Returns a SpanContext.
>     context: () => context,
> 
>     // By default does nothing
>     setAttribute: (_key: string, _value: unknown) => noopSpan,
> 
>     // By default does nothing
>     setAttributes: (_attributes: SpanAttributes) => noopSpan,
> 
>     // By default does nothing
>     addEvent: (_name: string, _attributes?: SpanAttributes) => noopSpan,
> 
>     // By default does nothing
>     setStatus: (_status: SpanStatus) => noopSpan,
> 
>     // By default does nothing
>     updateName: (_name: string) => noopSpan,
> 
>     // By default does nothing
>     end: (_endTime?: TimeInput) => {},
> 
>     // isRecording always returns false for noopSpan.
>     isRecording: () => false,
> 
>     // By default does nothing
>     recordException: (_exception: Exception, _time?: TimeInput) => {}
>   };
> 
>   return noopSpan;
> }
> ```
> 
> This to both explicit as it's creating a new span, but we can then hide the details within the helper and the NoopSpan class code just disappears -- further assisting minification.
> 
> And size wise this becomes (could be optimized a little more)
> 
> ```
> function createNoopSpan() {
>     var context = spancontext_utils_1.INVALID_SPAN_CONTEXT;
>     var noopSpan = {
>         // Returns a SpanContext.
>         context: function () { return context; },
>         // By default does nothing
>         setAttribute: function (_key, _value) { return noopSpan; },
>         // By default does nothing
>         setAttributes: function (_attributes) { return noopSpan; },
>         // By default does nothing
>         addEvent: function (_name, _attributes) { return noopSpan; },
>         // By default does nothing
>         setStatus: function (_status) { return noopSpan; },
>         // By default does nothing
>         updateName: function (_name) { return noopSpan; },
>         // By default does nothing
>         end: function (_endTime) { },
>         // isRecording always returns false for noopSpan.
>         isRecording: function () { return false; },
>         // By default does nothing
>         recordException: function (_exception, _time) { }
>     };
>     return noopSpan;
> }
> exports.createNoopSpan = createNoopSpan;
> ```
> 
> vs the following where all of the `NoopSpan.prototype.funcname` are not minified.
> 
> ```
> var NoopSpan = /** @class */ (function () {
>     function NoopSpan(_spanContext) {
>         if (_spanContext === void 0) { _spanContext = spancontext_utils_1.INVALID_SPAN_CONTEXT; }
>         this._spanContext = _spanContext;
>     }
>     // Returns a SpanContext.
>     NoopSpan.prototype.context = function () {
>         return this._spanContext;
>     };
>     // By default does nothing
>     NoopSpan.prototype.setAttribute = function (_key, _value) {
>         return this;
>     };
>     // By default does nothing
>     NoopSpan.prototype.setAttributes = function (_attributes) {
>         return this;
>     };
>     // By default does nothing
>     NoopSpan.prototype.addEvent = function (_name, _attributes) {
>         return this;
>     };
>     // By default does nothing
>     NoopSpan.prototype.setStatus = function (_status) {
>         return this;
>     };
>     // By default does nothing
>     NoopSpan.prototype.updateName = function (_name) {
>         return this;
>     };
>     // By default does nothing
>     NoopSpan.prototype.end = function (_endTime) { };
>     // isRecording always returns false for noopSpan.
>     NoopSpan.prototype.isRecording = function () {
>         return false;
>     };
>     // By default does nothing
>     NoopSpan.prototype.recordException = function (_exception, _time) { };
>     return NoopSpan;
> }());
> exports.NoopSpan = NoopSpan;
> ```

This will break instanceof checks. Not sure if anyone is using these checks at the moment but is a possibility worth considering.

MSNev commented:
> This will break instanceof checks. Not sure if anyone is using these checks at the moment but is a possibility worth considering.

Yes, it would and I see that in the tests, if this is really required (which I would think the idea of having a Noop version assumes that code doesn't want or need to check) then a possible "workaround" would be to check whether the span.context() === INVALID_SPAN_CONTEXT

Flarna commented:
Currently there is `class NoRecordingSpan extends NoopSpan` which would also not work anymore. I know https://github.com/open-telemetry/opentelemetry-js/pull/1900 removes this.

But I just remembered that [spec tells](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span-creation) that only a `Tracer` is allowed to create a span. So it seems a factory method on `Tracer` would be needed instead of exposing the `NoopSpan` construtor.

I assume this is again bad for minification but IMHO the ship to use classes,.. in OTel API has sailed. I don't think that changing `NoopSpan` from a `class` to something else but keep the rest of the API is a step forward.

Otel API is not yet GA, if proper minimizing is a major concern I recommend to start now but not on single, small classes instead rework the complete API.

Flarna commented:
> Yes, it would and I see that in the tests, if this is really required (which I would think the idea of having a Noop version assumes that code doesn't want or need to check) then a possible "workaround" would be to check whether the span.context() === INVALID_SPAN_CONTEXT

There are NoopSpans with valid context because of [this](https://github.com/open-telemetry/opentelemetry-specification/blob/v0.7.0/specification/trace/api.md#behavior-of-the-api-in-the-absence-of-an-installed-sdk), see https://github.com/open-telemetry/opentelemetry-js/blob/3b5892b5ad18e225759ab55d950ddaf74f3aa3a6/packages/opentelemetry-api/src/trace/NoopTracer.ts#L40

dyladan commented:
There are not too many real classes in the API. The API singletons and the noop classes I think is it. It probably wouldn't take too much effort to change if minification is a real concern. The SDK is another matter.

MSNev commented:
> There are not too many real classes in the API. The API singletons and the noop classes I think is it. It probably wouldn't take too much effort to change if minification is a real concern. The SDK is another matter.

Generally, minification concerns are not a single "big" thing, but more of death by a thousand cuts, so (apart from @Flarna [spec link](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span-creation)) requiring that everyone (SDK or otherwise) use ```new NoopSpan()``` vs a factory method that can get minified to ```m()``` from ```factoryMethod()``` is problematic.

But it seems that based on the spec we need to change this anyway so that only a tracer can create a span, so choosing the name of factory method on a tracer should be consideration as the length on the name has direct affects... i.e. ```createEmptyNoopSpan()``` vs ```noopSpan()``` or something similar (would depend on the usage model) etc. 

But to be spec compliant, it would seem that this functionality should be moved and (ideally) the NoopSpan class removed or not exported at least from the API.

As a workaround it is possible that *if* a particularity method is called multiple times in a block using a local variable can help -- I don't see that in this case though.

Flarna commented:
I looked once a again into this and it seems it's not that straight forward to solve this.
Actually there is already a factory method for `NoopSpan`: `NOOP_TRACER.startSpan()`.

There is a place where using `NOOP_TRACER.startSpan()` would look really weird as it looks like we start a span which is never ended: https://github.com/open-telemetry/opentelemetry-js/blob/7e423a9ec4975c28dc53e8c5d574685ade44f0e9/packages/opentelemetry-api/src/context/context.ts#L68

But as this is inside API it's possible to keep using `new NoopSpan()` there.

Flarna commented:
Update the PR to use 

But not sure if we should really go this path.
Looking into java it seems they reuse also a single invalid span and they expose an api to create such a span without using tracer.
see https://github.com/open-telemetry/opentelemetry-java/blob/faa8bf867a86a180cf4a4d5cdf2a04f83d82442b/api/all/src/main/java/io/opentelemetry/api/trace/Span.java#L58-L76

dyladan commented:
The prohibition on creating spans without a tracer applies primarily to users. Internally we can do what we think makes sense as long as it isn't leaked to the user. IMO we shouldn't export the NoopSpan from the API at all

MSNev commented:
> The prohibition on creating spans without a tracer applies primarily to users. Internally we can do what we think makes sense as long as it isn't leaked to the user. IMO we shouldn't export the NoopSpan from the API at all

I *looks* like we are only using NoopSpan from within the API (exception is the ot-tracing\test\Tracer.test.ts) so we *could* do the above and simple NOT export the NoopSpan from the Api -- thus making it private (in theory anyway) and the Api would be "*compliant*" from a users perspective (for NoopSpan)

dyladan commented:
NoopSpan is also used in the http plugin it looks like at least. This should probably be `tracer.startSpan('ignored', {}, suppressInstrumentation(ROOT_CONTEXT))`. I'll create another PR for this.

Flarna commented:
> NoopSpan is also used in the http plugin it looks like at least.

I removed all usages of `NoopSpan` and replaced it by `NOOP_TRACER.startSpan()`.

dyladan commented:
@open-telemetry/javascript-maintainers @open-telemetry/javascript-approvers please review PRs that affect the API. I want to merge what we have so I can move the code to the API repository this week.

vmarchaud commented:
>  This should probably be tracer.startSpan('ignored', {}, suppressInstrumentation(ROOT_CONTEXT)). I'll create another PR for this.

You closed the PR but this PR doesn't stop instrumentation right ? I think that was the goal with the non-recording span

dyladan commented:
> You closed the PR but this PR doesn't stop instrumentation right ? I think that was the goal with the non-recording span

Yes it does https://github.com/open-telemetry/opentelemetry-js/pull/1899/files#diff-26f74d8c9d46d751c084bc80c1cb73c74c4193f526d7ca8c5b4a0abb183b0287R590
